src/Core/Types.Selection/SingleOrDefaultMiddleware~1.cs

```diff
@@ -18,50 +17,40 @@ public SingleOrDefaultMiddleware(FieldDelegate next)
 
         public async Task InvokeAsync(IMiddlewareContext context)
         {
-            if (context.Field.ContextData[nameof(SingleOrDefaultOptions)] is
-                    SingleOrDefaultOptions options)
+            await _next(context).ConfigureAwait(false);
+
+            if (context.Result is IAsyncEnumerable<T> ae)
             {
-                await _next(context).ConfigureAwait(false);
+                await using IAsyncEnumerator<T> enumerator =
+                    ae.GetAsyncEnumerator(context.RequestAborted);
 
-                context.Result = context.Result switch
+                if (await enumerator.MoveNextAsync().ConfigureAwait(false))
                 {
-                    IAsyncEnumerable<T> ae => HandleAsyncEnumerable(options, ae),
-                    IEnumerable<T> ae => HandleEnumerable(context, options, ae),
-                    _ => context.Result
-                };
-            }
-        }
-
-        private async Task<T> HandleEnumerable(
-            IMiddlewareContext context,
-            SingleOrDefaultOptions options,
-            IEnumerable<T> e)
-        {
-            return await Task.Run(
-                    () => options.AllowMultipleResults ? e.FirstOrDefault() : e.SingleOrDefault(),
-                    context.RequestAborted)
-                .ConfigureAwait(false);
-        }
+                    context.Result = enumerator.Current;
+                }
+                else
+                {
+                    context.Result = default(T);
+                }
 
-        private async Task<T> HandleAsyncEnumerable(
-            SingleOrDefaultOptions options,
-            IAsyncEnumerable<T> ae)
-        {
-            T returnValue = default;
-            await foreach (T result in ae.ConfigureAwait(false))
+                if (await enumerator.MoveNextAsync().ConfigureAwait(false))
+                {
+                    context.Result = ErrorHelper.CreateMoreThanOneError(context);
+                }
+            }
+            else if (context.Result is IEnumerable<T> e)
             {
-                if (!Equals(returnValue, default(T)))
+                try
                 {
-                    throw new InvalidOperationException(
-                        "Sequence contains more than one element");
+                    context.Result = await Task.Run(
+                        () => e.SingleOrDefault(), context.RequestAborted)
+                        .ConfigureAwait(false);
                 }
-                returnValue = result;
-                if (options.AllowMultipleResults)
+                catch (InvalidOperationException)
```

Comment:
Should this be in `Task.Run` ?

Reply:
Yes probably better there since we than would not need to handle this async ...